src/webgldrawer.js

```diff
@@ -0,0 +1,1120 @@
+                        gl.bindBuffer(gl.ARRAY_BUFFER, this._firstPass.bufferOutputPosition);
+                        gl.vertexAttribPointer(this._firstPass.aOutputPosition, 2, gl.FLOAT, false, 0, 0);
+
+                        gl.bindBuffer(gl.ARRAY_BUFFER, this._firstPass.bufferTexturePosition);
+                        gl.vertexAttribPointer(this._firstPass.aTexturePosition, 2, gl.FLOAT, false, 0, 0);
+
+                        gl.bindBuffer(gl.ARRAY_BUFFER, this._firstPass.bufferIndex);
+                        gl.vertexAttribPointer(this._firstPass.aIndex, 1, gl.FLOAT, false, 0, 0);
+
+                        // Draw! 6 vertices per tile (2 triangles per rectangle)
+                        gl.drawArrays(gl.TRIANGLES, 0, 6 * numTilesToDraw );
+                    }
+                }
+
+                if(useTwoPassRendering){
+                    // Second rendering pass: Render the tiled image from the framebuffer into the back buffer
+                    gl.useProgram(this._secondPass.shaderProgram);
+
+                    // set the rendering target to the back buffer (null)
+                    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+
+                    // bind the rendered texture from the first pass to use during this second pass
+                    gl.activeTexture(gl.TEXTURE0);
+                    gl.bindTexture(gl.TEXTURE_2D, this._renderToTexture);
+
+                    // set opacity to the value for the current tiledImage
+                    this._gl.uniform1f(this._secondPass.uOpacityMultiplier, tiledImage.opacity);
+
+                    // bind buffers and set attributes before calling gl.drawArrays
+                    gl.bindBuffer(gl.ARRAY_BUFFER, this._secondPass.bufferTexturePosition);
+                    gl.vertexAttribPointer(this._secondPass.aTexturePosition, 2, gl.FLOAT, false, 0, 0);
+                    gl.bindBuffer(gl.ARRAY_BUFFER, this._secondPass.bufferOutputPosition);
+                    gl.vertexAttribPointer(this._firstPass.aOutputPosition, 2, gl.FLOAT, false, 0, 0);
+
+                    // Draw the quad (two triangles)
+                    gl.drawArrays(gl.TRIANGLES, 0, 6);
+
+                }
+
+                renderingBufferHasImageData = true;
+
+                if(useContext2dPipeline){
+                    // draw from the rendering canvas onto the output canvas, clipping/cropping if needed.
+                    this._applyContext2dPipeline(tiledImage, tilesToDraw, tiledImageIndex);
+                    renderingBufferHasImageData = false;
+                    // clear the buffer
+                    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+                    gl.clear(gl.COLOR_BUFFER_BIT); // clear the back buffer
+                }
+
+                // Fire tiled-image-drawn event.
+                // TODO: for discussion: do we need to ensure the image has been drawn to the output canvas already?
```

Comment:
Haven't thought about it in a while, but I could see it going a couple of ways:

1) Don't support this event for the `WebGLDrawer` but keep it as-is for `CanvasDrawer`. 
2) Change the event (possibly also the event name) to only fire once *all* `TiledImage`s have been drawn. What's the point of firing it after every image in a multi-image scenario, anyway? However, this would technically break an API.
3) Add a check to see whether there are any listeners bound to this event, and if so set `useContext2dPipeline = true` so that each `TiledImage` is drawn to the output canvas immediately, before the event is fired.

All of these options seem easy enough, but all have plusses and minuses. I'm happy to go with whatever you think would be best.

Reply:
Good points... I'll have to think about this. I suppose it doesn't need to block this PR.

I guess my initial thought is a combination of 1 and 2... Introduce a new event for when they've all been drawn and support that in all the drawers, but keep the current event in the older drawers, possibly with a deprecation notice.

That said, what kind of use is this event in the WebGL drawer? I suppose a plugin could also make draw calls on the same WebGL surface... Is that what we would expect the event to be used for?

Regarding the canvas pipeline, maybe that's something we'd want to be able to explicitly turn on (rather than implicitly by looking for event listeners)?

Reply:
And just to be clear: The reason we don't know if the image has been drawn yet is that we're batching the WebGL draws, right?

Reply:
Wait a second... `tiled-image-drawn` is being introduced by this PR, right? So there's no legacy use of it for us to worry about... We can change it to whatever makes sense. 

Also, if we just have it once, once they are all drawn, how is that different from `update-viewport`?

Reply:
Ah yes. I introduced this to try to keep the API for the new drawer as close as possible to the canvas drawer which fires the `tile-drawn` event - which really doesn't make sense for the webgl drawer. Adding the `tiled-image-drawn` event was a way to make the tests more uniform. We can totally get rid of it though, both here and in the canvas drawer (and associate tests). If you're amenable, we can make the tests just use `update-viewport` instead of `tile-drawn` I think (unless I'm missing something)

Reply:
Ok, with further testing, it seems that an event like this is needed for the test suite to work, because of how the tainted canvas checks are done. I had tried just using update-viewport and checking the status of the main canvas element but the tests all broke. I updated the description of the `tiled-image-drawn` event to make it clear this is used internally for testing and that `update-viewport` is to be used if you want to know when a frame has been drawn.

Reply:
I see! Good to know. Seems like it would be good to add a note that it's not exactly after each tiled image.

For that matter, would it make sense to switch it so it's after all the tiled images are drawn? That would still be more efficient, if it works.

Also, I don't think you need to have the JSDoc in both places... It'll just combine them somehow, or ignore one, since they are both the same event. You could document it in one place and leave a note in the other one about where the documentation is. Or you could move the event firing into a new function in DrawerBase and call it from both locations.

Reply:
I combined them into `DrawerBase` and marked it with `@private` as you suggested. I tried having it fire just once at the end of drawing (rather than once per image) but doing it that way requires keeping a reference to a `TiledImage` and `Tile` array outside of the loop for scope purposes, which seemed messy. I suppose I could only fire it after the first `TiledImage` is drawn rather than all of them, since it is only intended to be used by tests anyway. Let me know if you think this would be OK.

Reply:
Sure, if having it fire after the first works, that sounds good. I just figure we don't want to do extra work in the main loop if we don't have to.

Reply:
Added a check to only fire after the first one